### PR TITLE
Temperature Sensor plus Timing Improvements

### PR DESCRIPTION
**Pull request by hpmax**
This makes two changes to the current master.

1) It adds the temperature sensor.  Three additional variables are added to the config_file parser:

OUTPUT_POV_T (this should be pretty obvious)
temp_config (# of databits 9 through 12 -- integer)
temp_rate (How many updates per second would you like -- this is a floating point number)

Based on temp_rate, the code calculates the number of 12.5ms intervals between conversions.  The code will initiate a conversion and then poll every 12.5us to see if it is complete.  Once complete, it will report the value through the NMEA interface and then count the number of intervals until the pre-calculated amount.  If you already exceeded the number (i.e. you demanded a higher sampling rate than the DS18B20 could do), a new conversion will occur on the next interval.

2) The Kalman filter is provided a "delta_t" parameter which is set at 50ms.  The code is keeping track of time by counting the number of 12.5ms sleep cycles (ie 4 sleeps).  However, there is no guarantee that this time is accurate, and people have expressed concern.  This change records the clock time immediately prior to initiating a pressure sensor ADC conversion for all of the pressure sensors.  The Kalman filter is given the time difference between the prior two samples rather than 50ms.

### CODE REVIEW

File: main.c [hunk 1]
```diff
@@ -237,7 +241,66 @@ int NMEA_message_handler(int sock)
...
+			if (!done) {
+				if (OWReadByte(&temp_sensor)>0) {
+					if (OWReset(&temp_sensor)==1) {  // Reset
+						if (OWWriteByte(&temp_sensor,0xCC)==1) { // Skip ROM
+							if (OWWriteByte(&temp_sensor,0xBE)==1) { //Read Scratchpad
+								OWReadTemperature(&temp_sensor); // Convert output to temperature
+								if (temp_sensor.valid==1) {
+									done=1;
+									// Compose POV NMEA sentences
+									result=Compose_Temperature_POV(&s[0], temp_sensor.temperature);
```
Comment (linuxianer99) [inline main.c]: Please do not compose the NMEA sentence in the measurement handler routine.
Just sample the value, but compose the NMEA sentence in nmea_handler.

File: main.c [hunk 2]
```diff
@@ -237,7 +241,66 @@ int NMEA_message_handler(int sock)
...
+									// printf ("temp: %f\n",temp_sensor.temperature);
+									// NMEA sentence valid ?? Otherwise print some error !!
+									if (result != 1) printf("POV Temperature NMEA Result = %d\n",result);
+									// Send NMEA string via socket to XCSoar
+									// send complete sentence including terminating '\0'
+									if ((sock_err = send(sock, s, strlen(s)+1, 0)) < 0) fprintf(stderr, "send failed\n");
```
Comment (linuxianer99) [inline main.c]: The socket is not needed by measurement handler but only by the NMEA handler.

File: main.c [hunk 3]
```diff
@@ -320,7 +383,7 @@ void pressure_measurement_handler(void)
 			else
 			{
 				// of tep pressure
-				KalmanFiler1d_update(&vkf, tep_sensor.p/100, 0.25, 0.05);
+				KalmanFiler1d_update(&vkf, tep_sensor.p/100, 0.25, DELTA_TIME(tep_sensor.curtime,tep_sensor.prevtime));
```
Comment (linuxianer99) [inline main.c:325]: This change make sense, but please put this in a seperate commit as it nothing to do with temperature sensor support.
It is very difficult to track errors and do testing if different logic function are merged with one commit.

**Review by linuxianer99 — CHANGES_REQUESTED** [review]
Please have a look at my comments ...

**Comment by hpmax** [comment]
I removed all the temperature sensor stuff.  I am going to test it tomorrow and then do a commit.  Once you accept the commit, I'll re-introduce the temp sensor and do another commit.  I'd like to get this finished and move on with other stuff.

Regarding composing the sentences in NMEA handler... I understand your perspective.  However, the pressure_handler produces pressure values on reliable intervals, the temperature_handler doesn't.   I could have it check if there is a valid temperature reading available each time through and if so, compose the sentence.  If that's okay, I'll do that.

**Comment by linuxianer99** [comment]
Can you please rebase this ?
Otherwise we have lots of commits which make no sense ...

Thanks !

**Comment by hpmax** [comment]
Okay, this version is timing improvement only.  I want to look through some of the temperature data I collected before pushing the temperature improvements, but it should be coming soon.  As mentioned in Issue 23, there are a few other improvements/ideas I am looking at.  I'm also at some point going to try to re-do the Kalman filter.  It would be nice to perhaps use the IMU (either internal or external) or to use the static and pitot systems to supplement (or replace) this.

**Comment by hpmax** [comment]
I'm not a git expert.  You want me to delete my fork, re-fork, commit my code into it and then issue anothet PR?

**Comment by kedder** [comment]
> You want me to delete my fork, re-fork, commit my code into it and then issue anothet PR?

No, specifically, *never* do that. Use this [tutorial](https://hackernoon.com/beginners-guide-to-interactive-rebasing-346a3f9c3a6d) (or any other) to rebase/reorganize commits and the force-push to the same branch. That will update the PR.

**Comment by hpmax** [comment]
Okay, I think, its fixed.

**Comment by hpmax** [comment]
I'm going to close this and kill off this fork since my latest code is radically different.

**Comment by hpmax** [comment]
I deleted the repository and am starting over with my new timing improvement.  Once that is pushed, I will get either move back to the temperature sensor and/or the improved Kalman filter

**Pull request closed without merging**